### PR TITLE
Adds arkouda.testing module

### LINKED ISSUE #3143: add assert_series_equal function
Add an `assert_series_equal` function similar to pandas:
https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_series_equal.html

This will make testing easier.

### LINKED ISSUE #3144: add assert_index_equal function
Add an `assert_index_equal` function similar to pandas:
https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_index_equal.html

This will make testing easier.

### LINKED ISSUE #3145: add assert_frame_equal function
Add a `assert_frame_equal` function similar to pandas:
https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html

This will make testing easier.

### LINKED ISSUE #3212: Create testing library to mirror pandas
Create a testing library to mirror pandas:
https://github.com/pandas-dev/pandas/blob/main/pandas/_testing/asserters.py

Include `assert_frame_equal`, `assert_index_equal`, and `assert_series_equal`:
https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html
https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_index_equal.html
https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_series_equal.html

### PR DESCRIPTION
**Pull request by ajpotts**
This commit adds an `arkouda.testing` library to match `pandas.testing`.  It replicates code from the `asserters` module from pandas:
https://github.com/pandas-dev/pandas/blob/main/pandas/_testing/asserters.py

Including `assert_frame_equal`, `assert_index_equal`, and `assert_series_equal`:
https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html
https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_index_equal.html
https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_series_equal.html

Closes #3212 Create testing library to mirror pandas 
Closes #3145 add assert_frame_equal function  
Closes #3144 add assert_index_equal function  
Closes #3143 add assert_series_equal function

### CODE REVIEW

**Review by stress-tess — COMMENTED** [review]
A couple things to look at

Also I made sure to start with the actual `_asserter.py` file to give it my best review. But I gotta admit I was running out of a steam by the time I got to the `asserters_test`. Nothing popped out at me, but I think it would be good if one of the other reviewers checks that one out a bit more thoroughly

File: tests/groupby_test.py
[the remark this answers is not in the document]
Comment (ajpotts) [reply tests/groupby_test.py:1]: On this branch:
`-rw-rw-r--  1 amandapotts amandapotts 32910 Jun  7 14:38 groupby_test.py`

on master:
`-rwxrwxr-x 1 amandapotts amandapotts 32910 Jun 10 09:50 groupby_test.py`

File: arkouda/testing/_asserters.py [hunk 1]
```diff
@@ -0,0 +1,1057 @@
...
+    argsort,
+    array,
+    pdarray,
+    sort,
+    sum,
+)
+from arkouda.numpy import nan
+from arkouda.util import is_numeric
+
+DEBUG = True
```
[the remark this answers is not in the document]
Comment (ajpotts) [reply arkouda/testing/_asserters.py]: I meant to change this to `False`, thanks for reminding me.  Pandas squashes the stack trace for all the asserters, which makes sense in production, but is very annoying for developing/testing.  So I made it a flag that could be set to True.  We could get rid of it but I would kind of like to keep it for now if that's OK, b/c I still have some additions/changes to make to this module.

File: arkouda/testing/_asserters.py [hunk 2]
```diff
@@ -0,0 +1,1057 @@
...
+
+    if not isinstance(left, cls):
+        raise AssertionError(f"{cls_name} Expected type {cls}, found {type(left)} instead")
+    if not isinstance(right, cls):
+        raise AssertionError(f"{cls_name} Expected type {cls}, found {type(right)} instead")
+
+
+def assert_dict_equal(left, right, compare_keys: bool = True) -> None:
+    """
+    Asset that two dictionaries are equal.
```
[the remark this answers is not in the document]
Comment (ajpotts) [reply arkouda/testing/_asserters.py]: Fixed.  Thanks.

File: arkouda/testing/_asserters.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stress-tess) [reply arkouda/testing/_asserters.py]: yah I'm totally fine with keeping it! And I think when seeing a debug flag set to false by default, it's pretty intuitive that you should flip it to true when debugging

File: arkouda/testing/_asserters.py [hunk 3]
```diff
@@ -0,0 +1,1057 @@
...
+
+    if left_attr is right_attr:
+        return None
+
+    try:
+        result = left_attr == right_attr
+    except TypeError:
+        result = False
+    if (left_attr is nan) ^ (right_attr is nan):
+        result = False
```
[the remark this answers is not in the document]
Comment (ajpotts) [reply arkouda/testing/_asserters.py:379]: I'm trying to match pandas as exactly as possible given the current state of alignment.  This is the line I was trying to replicate:
```
    if (left_attr is pd.NA) ^ (right_attr is pd.NA):
        result = False
```
pd.NA covers np.nan and None.  So given you're comment, I'm changing to this instead:
```
    if (left_attr is None) ^ (right_attr is None):
        result = False
```
Let me know what you think.

File: arkouda/testing/_asserters.py [hunk 4]
```diff
@@ -0,0 +1,1057 @@
...
+    left,
+    right,
+    check_dtype: bool = True,
+    err_msg=None,
+    check_same=None,
+    obj: str = "pdarray",
+    index_values=None,
+) -> None:
+    """
+    Check that 'ak.pdarray' is equivalent.
```
[the remark this answers is not in the document]
Comment (ajpotts) [reply arkouda/testing/_asserters.py]: I was trying to match pandas:

```
def assert_numpy_array_equal(
    left,
    right,
    strict_nan: bool = False,
    check_dtype: bool | Literal["equiv"] = True,
    err_msg=None,
    check_same=None,
    obj: str = "numpy array",
    index_values=None,
) -> None:
    """
    Check that 'np.ndarray' is equivalent.

    Parameters
    ----------

```

But I don't think changing the docstring is a big deal so I made the change.

File: arkouda/testing/_asserters.py [hunk 5]
```diff
@@ -0,0 +1,1057 @@
...
+
+    if isinstance(left, pdarray):
+        left = left.to_ndarray()
+    if isinstance(left, np.ndarray):
+        left = pprint_thing(left)
+    elif isinstance(left, (Categorical, Strings)):
+        left = repr(left)
+
+    if isinstance(right, pdarray):
+        right = right.to_ndarray()
```
[the remark this answers is not in the document]
Comment (ajpotts) [reply arkouda/testing/_asserters.py:481]: I changed it to this which seems to give the correct string:

```
    if isinstance(left, np.ndarray):
        left = pprint_thing(left)
    elif isinstance(left, (Categorical, Strings, pdarray)):
        left = repr(left)
```

File: arkouda/testing/_asserters.py [hunk 6]
```diff
@@ -0,0 +1,1057 @@
...
+        raise_assert_detail(obj, msg, left_attr, right_attr)
+    return None
+
+
+def assert_is_sorted(seq) -> None:
+    """Assert that the sequence is sorted."""
+    if isinstance(seq, (Index, Series)):
+        seq = seq.values
+    # sorting does not change precisions
+    assert_arkouda_array_equal(seq, sort(array(seq)))
```
[the remark this answers is not in the document]
Comment (ajpotts) [reply arkouda/testing/_asserters.py:394]: I could make the change, but I'm not sure it's worth it...  I'm trying to match pandas as exactly as I can.  

This is the code I'm trying to match:
```
def assert_is_sorted(seq) -> None:
    """Assert that the sequence is sorted."""
    if isinstance(seq, (Index, Series)):
        seq = seq.values
    # sorting does not change precisions
    if isinstance(seq, np.ndarray):
        assert_numpy_array_equal(seq, np.sort(np.array(seq)))
    else:
        assert_extension_array_equal(seq, seq[seq.argsort()])
```

I agree what you suggest would be more efficient, but I think the asserters will mainly be used for testing...

File: arkouda/testing/_asserters.py [hunk 7]
```diff
@@ -0,0 +1,1057 @@
...
+            if isinstance(left, pdarray) or isinstance(right, pdarray):
+                obj = "pdarray"
+            else:
+                obj = "Input"
+            assert_class_equal(left, right, obj=obj)
+
+        if isinstance(left, pdarray) and isinstance(right, pdarray):
+            assert np.allclose(
+                left.to_ndarray(), right.to_ndarray(), rtol=rtol, atol=atol, equal_nan=True
+            )
```
[the remark this answers is not in the document]
Comment (ajpotts) [reply arkouda/testing/_asserters.py:122]: Warning added.

File: tests/groupby_test.py
[the remark this answers is not in the document]
Comment (ajpotts) [reply tests/groupby_test.py:1]: Will be resolved by:
https://github.com/Bears-R-Us/arkouda/pull/3309

File: arkouda/testing/_asserters.py [hunk 8]
```diff
@@ -0,0 +1,1057 @@
...
+        # We want to skip exact dtype checking when `check_categorical`
+        # is False. We'll still raise if only one is a `Categorical`,
+        # regardless of `check_categorical`
+        if isinstance(left, Categorical) and isinstance(right, Categorical) and not check_categorical:
+            pass
+        else:
+            assert_attr_equal("dtype", left, right, obj=f"Attributes of {obj}")
+
+    if check_exact or not is_numeric(left.values) or not is_numeric(right.values):
+        assert_arkouda_array_equal(
```
[the remark this answers is not in the document]
Comment (ajpotts) [reply arkouda/testing/_asserters.py:887]: Added.

File: arkouda/testing/_asserters.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (stress-tess) [reply arkouda/testing/_asserters.py]: ah that makes sense! yah this sounds good to me

File: arkouda/testing/_asserters.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (stress-tess) [reply arkouda/testing/_asserters.py:401]: yah that makes sense to me!

File: arkouda/testing/_asserters.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (stress-tess) [reply arkouda/testing/_asserters.py]: ah okay! that makes sense

File: arkouda/testing/_asserters.py [hunk 9]
```diff
@@ -0,0 +1,1057 @@
...
+    DataFrame,
+    Index,
+    MultiIndex,
+    Series,
+    Strings,
+    argsort,
+    array,
+    pdarray,
+    sort,
+    sum,
```
[the remark this answers is not in the document]
Comment (stress-tess) [reply arkouda/testing/_asserters.py]: yah that makes since

**Review by stress-tess — APPROVED** [review]

**Comment by stress-tess** [comment]
@ajpotts this is good to go other than some changes I think from the dataframe indexing revert: https://github.com/Bears-R-Us/arkouda/pull/3323

**Review by jaketrookman — APPROVED** [review]
Looks good

**Review by stress-tess — APPROVED** [review]

**Pull request merged**